### DISCUSSION TITLE
J2000 to Now - Rotation_EQJ_EQD

### DISCUSSION THREAD

**Discussion by mikasmart** [cosinekitty/astronomy#114] [Q&A] [answered]
Hi !

Thank you for this wonderful module!

I am looking for a way to get the position of a star on today's date. I have the coordinates of the star but in J2000 format (ra/dec).
I view a _Rotation_EQJ_EQD_ function but I can't figure out how to use it and there is no demo about it.

Someone can help me ? My code is in javascript.

**Comment by cosinekitty** [comment, marked as answer]
Hi @mikasmart ! That is a great question. You were on the right track looking at `Rotation_EQJ_EQD`. This is not a trivial conversion, so I have added a demo to illustrate it for you.  Take a look at [equator_of_date.js](https://github.com/cosinekitty/astronomy/blob/master/demo/nodejs/equator_of_date.js).  Here is a sample of how to use it for the equator at the current date and time:

```
$ node equator_of_date.js 12 45
time = 2021-07-07T00:56:06.051Z
Equator-of-date coordinates: RA=12.0181, DEC=44.8819
```

Or you can optionally specify a particular date and time to use:

```
$ node equator_of_date.js 12 45 1972-02-14T14:35:00Z
time = 1972-02-14T14:35:00.000Z
Equator-of-date coordinates: RA=11.9765, DEC=45.1535
```

**Reply by mikasmart** [reply]
Thanks ! It's exactly what i mean.

**Comment by mikasmart** [comment]
I did a test with M57 and using data from Stelarium. The results are very close but incorrect.
Maybe the calculation lacks precision?
	
With M57 (data from Stelarium)
Date	/  Degrees / Decimal conversion
J2000	/ 18h53m35.08 - 33°01'45.0"	/ 18.89308 - 33.02917
NOW	/ 18h54m23.72 - 33°03'24.4"	/ 18.90659 - 33.05678

Result from equator_of_date (Must be equal to M57 & NOW)

NOW / 18h54m22.32 - 33°03'21.6" / 18.90621 - 33.05597

Test started with command `node equator_of_date 18.89308 33.02917`

**Reply by cosinekitty** [reply]
That is an interesting test. I was able to reproduce this discrepancy with M57 as you did. Also I tried the star Procyon. Because Procyon is a point source instead of a diffuse object, there would be no doubt about its exact J2000 coordinates. Then I used the [web version of Stellarium](https://stellarium-web.org/).

Here is what I believe is happening. Astronomy Engine corrects for both precession and nutation of the Earth's axis. It appears that Stellarium is correcting only for precession. I made [this Google speadsheet](https://docs.google.com/spreadsheets/d/1Dmx88BfN_DbzrH6XZtThyPH0100asdB1FLM-QLvF85M/edit?usp=sharing) so you can follow along with my calculations.

When I leave Astronomy Engine as-is, cells D8 and D9 in the spreadsheet show the discrepancy in arcminutes between Stellarium and Astronomy Engine. This is a large discrepancy: nearly one arcminute!  However, when I edit `astronomy.js` and change this:

```JavaScript
function Rotation_EQJ_EQD(time) {
    time = MakeTime(time);
    const prec = precession_rot(time, PrecessDirection.From2000);
    const nut = nutation_rot(time, PrecessDirection.From2000);
    return CombineRotation(prec, nut);
}
```

to this (removing the nutation correction):

```JavaScript
function Rotation_EQJ_EQD(time) {
    time = MakeTime(time);
    const prec = precession_rot(time, PrecessDirection.From2000);
    return prec;
    //const nut = nutation_rot(time, PrecessDirection.From2000);
    //return CombineRotation(prec, nut);
}
```

The discrepancy becomes tiny. See spreadsheet cells F8 and F9. The error has decreased to less than 1 arcsecond, which is about what I would expect, because of some simplifications Astronomy Engine makes to be fast and small compared to other astronomy algorithms. AE treats radio-based [ICRF](https://en.wikipedia.org/wiki/International_Celestial_Reference_Frame) and older optical-based J2000 orientation systems as interchangeable.

I have verified Astronomy Engine's precession and nutation algorithms against the NOVAS C 3.1 code they are based on. I use a truncated version of the NOVAS series, and I estimated their difference would never be greater than about 2.2 arcseconds.

I'm not sure why Stellarium does not include nutation in their conversions, but it is probably because the difference does not matter much for amateur astronomers, and it certainly would make their code run faster for maintaining good animation frame rates.

**Reply by cosinekitty** [reply]
Oh no, this is embarrassing! I realize I made a mistake. There is still a large discrepancy. I need to investigate more. Please stand by...

**Reply by cosinekitty** [reply]
I can see now that Stellarium *does* perform both precession and nutation calculations. I made a silly mistake in my spreadsheet that fooled me into thinking that it didn't. I have been looking at their source code, both for the web version and their native C++ version. I'm still trying to figure out the discrepancy. Here are some of things I'm looking at:

- The simple example I provided does not correct for aberration. I'm not sure what your goals are, so aberration may or may not be a helpful correction. This can account for a position difference of more than 20 arcminutes.
- Nearby stars (like Procyon) can have significant proper motion over the time span from 2000 to 2021. I doubt a distant object like M57 would be affected much though. But for Procyon, this changes its position by -0.25 arcminutes in RA and -0.37 arcminutes in DEC. In general, one needs to adjust for the movement of stars over time.
- I intend to add more rigorous testing of precession and nutation calculations. Perhaps I will use Stellarium source code to generate some test data and verify it against Astronomy Engine calculations. I'm pretty sure I would have noticed such a large error by now, but I want to make sure there is nothing wrong with this part of the code.

@mikasmart, can you tell me more about what you are trying to do? It would help me figure out whether aberration corrections are a good idea for your project.

**Reply by mikasmart** [reply]
> @mikasmart, can you tell me more about what you are trying to do? It would help me figure out whether aberration corrections are a good idea for your project.

With pleasure! I would very much like to discuss my project but I don't want to talk about it here. My project uses Arduino & smartphone.
My email address: (redacted) - Contact me for more information

**Reply by cosinekitty** [reply]
@mikasmart, I am working on new functions in Astronomy Engine to help correct for stellar aberration, which can cause errors of up to 20.5 arcseconds (about 1/3 of an arcminute). I believe this is part of the discrepancy you discovered between Stellarium and Astronomy Engine.

I estimate this work will take about a week to finish. Then I will be in a better position to repeat tests with Stellarium and see if I'm getting closer to agreement with it.

I will update here when I have something interesting to report.

**Reply by cosinekitty** [reply]
@mikasmart I just updated the `equator_of_date.js` to correct for aberration, and the results are very encouraging. Before, for the M57 Ring Nebula, the discrepancy between Stellarium and Astronomy Engine was about 0.4 arcminutes. After correcting for aberration the disagreement is less than 0.01 arcminutes. I hope you can try the latest version and confirm that it is more in agreement with Stellarium.

**Comment by laheller** [comment]
Hi @cosinekitty 

First of all, let me congratulate you on your astronomy engine library! I just found it and still wondering, how well it is implemented and how easy is to use it! I am coming from the C#.NET world and I also tried other libraries, like for example Joe Sauve's [AASharp](https://github.com/jsauve/AASharp) which is very good too but has somewhat poor or almost zero documentation.

Nevemind, I had the same question as @mikasmart and meanwhile I am thinking about, whether is it possible to do it a different way, using the [Astonomy.Equator( )](https://github.com/cosinekitty/astronomy/tree/master/source/csharp#Astronomy.Equator) method, like this:

I have one star from the [Henry Draper catalog](https://vizier.cds.unistra.fr/viz-bin/VizieR?-source=III/135A), having its __RA.icrs_ and __DE.icrs_ coordinates, both are J2000 in decimal degrees.

My C# code is then:
```csharp
using CosineKitty;

// J2000 equatorial coordinates of the HD11150 star
var RA = 28.9986; // according to VizieR catalog desc it's in decimal degrees, later we need to convert to sidereal hours by dividing 15
var DE = 80.9102; // decimal degrees

var obs = new Observer(lat, lon, height);
var now = DateTime.UtcNow; // the date we want to convert to
var astro_now = new AstroTime(now);

// do the conversion
// first let's define a star, but switch RA coordinate in decimal degrees to sidereal hours, because the method expects it
Astronomy.DefineStar(Body.Star1, RA / 15.0, DE, 1000.0);
var coord = Astronomy.Equator(Body.Star1, astro_now, obs, EquatorEpoch.OfDate, Aberration.None);
// do something with the coord, e. g. display it
```
My question is, whether is the above calculation correct for the input star? Does the _Equator_ method convert J2000 to JNow coordinates this whay?

BR,

Ladislav

**Reply by cosinekitty** [reply]
Yes, using `EquatorEpoch.OfDate` will report the star's position using the Earth's equator at the time `astro_now`, correcting for precession and nutation.

**Reply by laheller** [reply]
Thank you!
